Kernel/Bus/PCI/Controller/HostController.cpp

```diff
@@ -128,10 +148,120 @@ UNMAP_AFTER_INIT void HostController::enumerate_attached_devices(Function<void(E
+{
+    // First, Assign PCI-to-PCI bridge bus numbering
+    u8 bus_id = 0;
+    enumerate_attached_devices([this, &bus_id](EnumerableDeviceIdentifier const& device_identifier) {
+        // called for each PCI device encountered (including bridges)
+        if (read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::CLASS) != PCI::ClassID::Bridge)
+            return;
+        if (read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBCLASS) != PCI::Bridge::SubclassID::PCI_TO_PCI)
+            return;
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SECONDARY_BUS, ++bus_id);
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBORDINATE_BUS, 0xFF); }, [this, &bus_id](EnumerableDeviceIdentifier const& device_identifier) {
+        // called after a bridge was recursively enumerated
+        write8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::SUBORDINATE_BUS, bus_id); });
+
+    // Second, Assign BAR addresses
+    // TODO: We currently naively assign addresses bump-allocator style - Switch to a proper allocator if this is not good enough
+    enumerate_attached_devices([this, &mmio_32bit_base, mmio_32bit_end, &mmio_64bit_base, mmio_64bit_end](EnumerableDeviceIdentifier const& device_identifier) {
+        // device-generic handling
+        auto header_type = read8_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), PCI::RegisterOffset::HEADER_TYPE);
+        auto const max_bar = (header_type == 0) ? RegisterOffset::BAR5 : RegisterOffset::BAR1;
+        for (u32 bar_offset = to_underlying(RegisterOffset::BAR0); bar_offset <= to_underlying(max_bar); bar_offset += 4) {
+            auto bar_value = read32_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), bar_offset);
+            auto bar_space_type = get_BAR_space_type(bar_value);
+            auto bar_prefetchable = (bar_value >> 3) & 1;
+            if (bar_space_type != BARSpaceType::Memory32BitSpace && bar_space_type != BARSpaceType::Memory64BitSpace)
+                continue; // We only support memory-mapped BAR configuration at the moment
+            if (bar_space_type == BARSpaceType::Memory32BitSpace) {
+                write32_field(device_identifier.address().bus(), device_identifier.address().device(), device_identifier.address().function(), bar_offset, 0xFFFFFFFF);
```

Comment:
I assume you are not using PCI::get_BAR_space_size here, because it would try to take a lock?

Reply:
PCI::get_BAR_space_size requires constructing a DeviceIdentifier (not the EnumerableDeviceIdentifier which I have), which I thought was wasteful.